### PR TITLE
[lit] RTL simulation script with Verilator and Questa support

### PR DESCRIPTION
**Pull request by teqdruid**
- Adds a python script (`circt-rtl-sim.py`) to compile and run RTL simulations in Verilator and Questa.
- Adds lit support for script.
- Adds cmake and lit support for Questa.
- Moves an existing test over.
- Standardizes top module name to `top`. (It was `main` in some cases.)

Will be used for #322.

### CODE REVIEW

**Review by darthscsi — APPROVED** [review]
LGTM

**Comment by lattner** [comment]
Is it possible to build an abstraction layer so these tests are simulator independent?  I'd much rather have tests be something like "REQUIRES: simulator" and then run a tool/bash script like "utils/circt-sim" that then forwards to VCS, questa, verilator etc to do the execution.

**Comment by teqdruid** [comment]
Certainly, though I'm not certain the complexity since I'm only familiar with Questa. Other tools may have different compilation / running flows which ties into the interface said tool would expose. I was thinking the same, but since I'm not familiar with other simulator's work flows was uncertain what this abstraction should minimally expose. Starting with questa-style and then incrementally developing it (by/with someone more familiar with simulation workflows than I) seems to be the right way to go. Also, there _could theoretically_ end up being questa-specific tests so that lit feature could come in handy.

I've spent the last few years working on an internal tool which has similar functionality but it's waaayyy more complex than we have need for. It has a declarative YAML file specification, which I suspect is overkill for our needs.

**Comment by teqdruid** [comment]
Actually, I take some of that back... I was thinking that verilator and commercial simulators were different enough that we couldn't use the same compilation script to drive both; however, the compilations are similar enough the we could probably. The functionality, however, is different enough that it may or may not make sense to. And we'd definitely need separate lit features.

**Comment by teqdruid** [comment]
@lattner I wrote a python script which abstracts out the simulator... I don't think it was worth it for Verilator to be mixed in (given the number of arguments it requires) but now that I've done it, it'll be some help. Assuming that other commercial simulators work similarly, it could ease their integration.

**Comment by lattner** [comment]
My sense is that all the simulators have the same core feature set of "run verilog" but then all have a long list of special and weird extensions.  Most of our integration testing will be about that core feature set, so lets start by nailing that.

If we end up needing (as part of the testsuite) some of the weird things, then we can add new features that only work in some situations.  That's ok, they may not even be simulator dependent, they may be simulator version dependent.  This is what cmake checks are good for.

File: integration_test/EmitVerilog/basic.mlir [hunk 1]
```diff
@@ -1,12 +1,11 @@
 // REQUIRES: verilator
 // RUN: circt-translate %s -emit-verilog -verify-diagnostics > %t1.sv
-// RUN: verilator --cc --top-module main -sv %t1.sv --build --exe %S/../driver.cpp
-// RUN: ./obj_dir/Vmain --cycles 8 2>&1 | FileCheck %s
+// RUN: circt-rtl-sim.py verilator %t1.sv %S/../driver.cpp --simargs "--cycles 8" 2>&1 | FileCheck %s
```
Comment (lattner) [inline integration_test/EmitVerilog/basic.mlir:3]: Instead of pushing the choice of simulator into the test, can you push it into circt-rtl-sim?  Rationale: There is nothing verilator specific about this test, that is just an implementation detail.  I'd prefer to see something like:

```
// REQUIRES: rtl-sim
// RUN: circt-translate %s -emit-verilog -verify-diagnostics > %t1.sv
// RUN: circt-rtl-sim.py --top-module top %t1.sv --cycles 8

or something like that.  the fact that verilator poops out a C++ file etc should all be handled within the script.  common flags like "--cycles 8" that can be supported by multiple simulators can then be handled by the script.

To put it another way, I'd love for this script to be the [`lli`](https://llvm.org/docs/CommandGuide/lli.html) of system verilog.

File: utils/circt-rtl-sim.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
```
Comment (lattner) [inline utils/circt-rtl-sim.py:1]: Please add file header boilerplate.

Also, I think it would make sense to name this `circt-sv-sim.py` since it is likely to be verilog specific, not arbitrary RTL.  If we ever add VHDL, that can be its own tool/script.

File: integration_test/EmitVerilog/basic.mlir (hunk 1 above)
Comment (teqdruid) [reply integration_test/EmitVerilog/basic.mlir:3]: That would be difficult and not terribly useful. Or very, very, extremely hard and very useful.

> Rationale: There is nothing verilator specific about this test, that is just an implementation detail.

Actually, there is. The `driver.cpp` program (which is part of the test) relies on several characteristics: the top level is specified as top, and there are `clk` and `rstn` pins which need to be driven appropriately. Testbenches which aren't written to target verilator would not expose any ports in their top module, so the produced SystemVerilog is intended for Verilator. They also don't need a C++ driver program.

This is a sort of contrived example as it is incredibly uninteresting. It just runs the produced SystemVerilog toggling the `clk` and `rstn` ports. Real tests would want to expose more I/O ports to test functionality. If using verilator, they'd have to write a custom C++ driver app (which could presumably control the number of cycles). If using a commercial simulator, the top level wouldn't expose any ports and it would drive and read internal wires via simulation constructs which verilator doesn't support. Would it be possible to automatically synthesize a C++ driver app with the same functionality? Yes, but said functionality isn't terribly interesting so it wouldn't be used much.

Bottom line: every single test today is written to target either verilator or a commercial simulator as they are _very_ different beasts. Should we do interesting things to advance the status quo? Absolutely, though that is out-of-scope for this script and would have to be thought through further. Should we create a 'simulation' dialect which can lower to either a C++ driver (for verilator) or commercial simulation constructs? Maybe. Should we require all testing be done via synthesizable SystemVerilog (so that it's supported by verilator and commercial sims)? Probably not.

I fully expect that for the foreseeable future that for anything interesting, it would have to target either Verilator or a commercial simulator.

If you look at the example I put in the description, you'll see that I did add footing for selecting a commercial simulator and using it. There your assumption that there is nothing simulator-specific applies. It wouldn't be too much harder to pass the default simulator to the script via an environment variable (to eliminate the %defaultSim) so I'll do that.

File: utils/circt-rtl-sim.py (hunk 2 above)
Comment (teqdruid) [reply utils/circt-rtl-sim.py:1]: Actually, mixed language simulations (SystemVerilog and VHDL) are common and very similar. In questa, it's just `vcom` instead of `vlog` so I think it makes sense to support VHDL with the same script.

File: integration_test/EmitVerilog/basic.mlir (hunk 1 above)
Comment (teqdruid) [reply integration_test/EmitVerilog/basic.mlir:3]: Caveat: Running an `uninteresting` simulation (only `clk` and `rst`) becomes interesting when paired with Cosim to drive it from another process. So I'll probably add test driver shell generation (C++ or SV) at some point in the future to support ESI cosim simulations.

Comment (lattner) [reply integration_test/EmitVerilog/basic.mlir:3]: Ok, I understand what you're saying here, but I still think you're missing my point.  :-)

We're likely to want many _simple_ tests for basic functionality that are not doing fancy things with clocks and reset logic.  We can have standardized ".cpp" file for verilator that works with "the one true way" and we can have a standardized module wrapper that gets thrown into vcs/questa or whatever.   We can just require all these style of tests to define their top level module with the same name like "top" - no need to get fancy here.

That said, this doesn't eliminate the utility of having a questa/verilator specific driver script for doing questa/verilator specific things.  It is just that simulator independent tests like this would have to be duplicated, and that will make me grumpy every time it happens :)

Investing in test infra like this pays off big time in the long run, because getting it wrong up front creates a lot of the "wrong thing" that has to be reworked later.

Comment (teqdruid) [reply integration_test/EmitVerilog/basic.mlir]: I added an SystemVerilog driver which has the option to kill the simulation after N cycles, just like driver.cpp. This allowed me to run basic_run_test.sv in both verilator and questa without changing it. I'm not confident that this'll be useful for anything other than self-stimulating tests and uninteresting tests, but time will tell.

Still have some cleanup to do. Just pushing for the night.

File: integration_test/basic_run_test.sv [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// RUN: circt-rtl-sim.py --cycles 2 %s
```
Comment (lattner) [inline integration_test/basic_run_test.sv:1]: Nice, I'd recommend naming this "circt-rtl-sim-basic.sv" or something like that, since it is testing the script.

File: integration_test/basic_run_test.sv [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+// This also works if you have Questa available:
+//   circt-rtl-sim.py --sim %comsim --cycles 2 %s
```
Comment (lattner) [inline integration_test/basic_run_test.sv:3]: Does this test need a REQUIRES: line?

Comment (lattner) [reply integration_test/basic_run_test.sv:3]: Also should it pipe the output into filecheck and have some CHECK lines?

File: integration_test/driver.cpp [hunk 5]
```diff
@@ -5,7 +5,7 @@
 //
 //===----------------------------------------------------------------------===//
 
-#include "Vmain.h"
+#include "Vtop.h"
```
Comment (lattner) [inline integration_test/driver.cpp:8]: Given that you have code that needs to be built, it might make sense to move circt-rtl-sim.py (and the .cpp and .sv file) into circt/tools/circt-rtl-sim/.  This would make sure the .o file for the .cpp file is always built before tests are run, simplifying the python file.

Comment (lattner) [reply integration_test/driver.cpp:8]: This would also be a more logical place to put testing utilities like this than in integration-test itself.

File: integration_test/driver.sv [hunk 6]
```diff
@@ -0,0 +1,61 @@
+//===- driver.sv - Standard SystemVerilog testbench driver ------*- C++ -*-===//
```
Comment (lattner) [inline integration_test/driver.sv:1]: *- C++ -*.   -> verilog, or just drop it

File: integration_test/driver.sv [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the top module driver for most simulations. It makes the following
+// assumptions/demands:
+//
+// - The testbench module is called 'top'.
+// - It exposes a pin named 'clk' (for the clock).
+// - It exposes a pin named 'rstn' (for the reset signel).
```
Comment (lattner) [inline integration_test/driver.sv:14]: signel -> signal

File: integration_test/driver.sv [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+// Make sure this file compiles.
+// REQUIRES: comsim
+// RUN: circt-rtl-sim.py --sim %comsim --no-run --no-default-driver %s
```
Comment (lattner) [inline integration_test/driver.sv:20]: I'm confused - is this a driver that is part of the runner, or is this a test?  If it is both, they should be split apart.

**Review by lattner — COMMENTED** [review]
Cool, this is looking like a good approach, thank you for pushing on this!

File: integration_test/driver.cpp (hunk 5 above)
Comment (teqdruid) [reply integration_test/driver.cpp:8]: This gets compiled along with the code which Verilator produces by Verilator so we don't compile it as part of the cmake build. I agree about the location.

File: integration_test/driver.sv (hunk 8 above)
Comment (teqdruid) [reply integration_test/driver.sv:20]: It's a sanity check for the driver itself. Since I'll be moving it, this is no longer appropriate.

File: integration_test/circt-rtl-sim/circt-rtl-sim-comsim.sv [hunk 9]
```diff
@@ -0,0 +1,15 @@
+// REQUIRES: comsim
```
Comment (lattner) [inline integration_test/circt-rtl-sim/circt-rtl-sim-comsim.sv:1]: Are you intentionally forking this test?  Would it be possible to have one test and add a new `REQUIRES: rtl-sim` flag that is enabled when either questa or verilator are available?

**Review by lattner — APPROVED** [review]
Looks really great, thank you again for driving this!

File: integration_test/circt-rtl-sim/circt-rtl-sim-comsim.sv (hunk 9 above)
Comment (teqdruid) [reply integration_test/circt-rtl-sim/circt-rtl-sim-comsim.sv:1]: In the case where both verilator and a commercial simulator are installed, I'd like both driver.cpp and driver.sv to be tested. Adding a feature which indicates that a simulator is available is a good idea.

Comment (lattner) [reply integration_test/circt-rtl-sim/circt-rtl-sim-comsim.sv:1]: makes sense, thanks!
